File: backend/model/menu.go
package model

import (
	"time"

	"github.com/lib/pq"
)

// Menu 菜单模型
type Menu struct {
	ID        uint `gorm:"primarykey"`
	CreatedAt time.Time
	UpdatedAt time.Time

	Name      string `gorm:"uniqueIndex;size:100;not null;comment:菜单唯一标识"`
	Path      string `gorm:"size:255;not null;comment:菜单路径"`
	Component string `gorm:"size:255;comment:组件路径"`
	ParentID  *uint  `gorm:"index;comment:父菜单ID"`
	Order     int    `gorm:"default:0;comment:排序"`

	// JSONB 字段，存储 meta 信息（title, icon, i18nKey 等）
	Meta string `gorm:"type:jsonb;comment:菜单元数据"`

	// 需要的角色列表
	RequiredRoles pq.StringArray `gorm:"type:text[];not null;default:'{}';comment:所需角色"`

	// 关联关系
	Parent   *Menu  `gorm:"foreignKey:ParentID"`
	Children []Menu `gorm:"foreignKey:ParentID"`
}

func (Menu) TableName() string {
	return "menus"
}

File: frontend/apps/src/api/system/menu.ts
import type { RequestClientConfig } from '@vben/request';

import { requestClient } from '#/api/request';

import { listOf } from '../_utils';

export namespace MenuApi {
  export interface RouteMeta {
    hideInMenu?: boolean;
    i18nKey?: string;
    icon?: string;
    localIcon?: string;
    order?: number;
    roles?: string[];
    title?: string;
  }

  export interface MenuItem {
    children?: MenuItem[];
    component: string;
    createdAt?: string;
    id: number;
    meta: RouteMeta;
    name: string;
    order: number;
    parentId: number;
    path: string;
    requiredRoles: string[];
  }

  export interface MenuListResult {
    list: MenuItem[];
  }

  export interface MenuPayload {
    component: string;
    meta: RouteMeta;
    name: string;
    order: number;
    parentId: number;
    path: string;
    requiredRoles: string[];
  }
}

export async function getMenuListApi(config?: RequestClientConfig) {
  const resp = await requestClient.get<MenuApi.MenuListResult>(
    '/menu/list',
    config,
  );
  return listOf(resp);
}

export async function createMenuApi(data: MenuApi.MenuPayload) {
  return requestClient.post<void>('/menu', data);
}

export async function updateMenuApi(id: number, data: MenuApi.MenuPayload) {
  return requestClient.put<void>(`/menu/${id}`, data);
}

export async function deleteMenuApi(id: number) {
  return requestClient.delete<void>(`/menu/${id}`);
}
